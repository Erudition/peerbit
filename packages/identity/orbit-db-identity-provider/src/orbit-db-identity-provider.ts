import { Identity } from './identity';
import { IdentityProvider } from './identity-provider-interface'
import { Keystore, SignKeyWithMeta } from '@dao-xyz/orbit-db-keystore'
import { Ed25519PublicKey } from 'sodium-plus';
import { verifySignatureEd25519 } from '@dao-xyz/peerbit-crypto';

export class OrbitDBIdentityProvider extends IdentityProvider {
  _keystore: Keystore;
  constructor(keystore) {
    super()
    if (!keystore) {
      throw new Error('OrbitDBIdentityProvider requires a keystore')
    }
    this._keystore = keystore
  }

  // Returns the type of the identity provider
  static get type() { return 'orbitdb' }

  async getId(options: { id?: Uint8Array } = {}) {
    const id = options.id
    if (!id) {
      throw new Error('id is required')
    }

    const keystore = this._keystore
    const idString = Buffer.from(id).toString('base64');
    const existingKey = await keystore.getKeyByPath(idString, SignKeyWithMeta);
    const key = (existingKey) || (await keystore.createKey(idString, SignKeyWithMeta))
    return new Uint8Array(key.publicKey.getBuffer());
  }

  async sign(data: string | Uint8Array | Buffer, options: { id?: Uint8Array } = {}) {
    const id = options.id
    if (!id) {
      throw new Error('id is required')
    }
    const keystore = this._keystore
    const idString = Buffer.from(id).toString('base64');
    const key = await keystore.getKeyByPath(idString, SignKeyWithMeta)
    if (!key) {
      throw new Error(`Signing key for '${idString}' not found`)
    }
    return Keystore.sign(data, key);
  }

  /* static async verify(signature: Uint8Array, data: Uint8Array, publicKey: Uint8Array): Promise<boolean> {
    return Keystore.verify(
      signature,
      new Ed25519PublicKey(Buffer.from(publicKey)),
      data,
    )
  } */

  static async verifyIdentity(identity: Identity) {
    // Verify that identity was signed by the ID
    if (identity.id instanceof Ed25519PublicKey) {
      return verifySignatureEd25519(identity.signatures.publicKey, identity.id, Buffer.concat([identity.publicKey.getBuffer(), identity.signatures.id]));
      /* return OrbitDBIdentityProvider.verify(
        identity.signatures.publicKey,
        new Uint8Array(Buffer.concat([identity.publicKey.getBuffer(), identity.signatures.id])),
        identity.id
      ) */
    }
    return false;
  }

}

