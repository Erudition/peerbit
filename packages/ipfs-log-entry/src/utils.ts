export const bigIntMax = (...args) => args.reduce((m, e) => e > m ? e : m);
export const bigIntMin = (...args) => args.reduce((m, e) => e < m ? e : m);
