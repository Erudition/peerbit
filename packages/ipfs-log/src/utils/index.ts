import { difference } from './difference'
import { findUniques } from './find-uniques'
import { isDefined } from './is-defined'
import { io } from 'orbit-db-io'

module.exports = {
  difference,
  findUniques,
  isDefined,
  io
}
