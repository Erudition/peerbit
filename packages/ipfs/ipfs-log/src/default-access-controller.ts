const EventEmitter = require('events').EventEmitter
import { MaybeEncrypted } from "@dao-xyz/peerbit-crypto";
import { Payload } from '@dao-xyz/ipfs-log-entry';
import { Ed25519PublicKey, PublicKey } from '@dao-xyz/peerbit-crypto'
import { Ed25519PublicKey } from 'sodium-plus'
// TODO extend IPFS-LOG access controller interface for canAppend method
/**
 * Interface for OrbitDB Access Controllers
 *
 * Any OrbitDB access controller needs to define and implement
 * the methods defined by the interface here.
 */
export interface CanAppendAccessController<T> {

  get allowAll(): boolean;
  canAppend(payload: MaybeEncrypted<Payload<T>>, key: MaybeEncrypted<PublicKey>): Promise<boolean>;

}

export class DefaultAccessController<T> implements CanAppendAccessController<T> {
  get allowAll(): boolean {
    return true;
  }

  async canAppend(payload: MaybeEncrypted<Payload<T>>, key: MaybeEncrypted<PublicKey>): Promise<boolean> {
    return true
  }
}
