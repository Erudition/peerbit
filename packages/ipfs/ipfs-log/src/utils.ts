// @ts-ignore
import isNode from 'is-node';

export const max = <T>(...args: T[]) => args.reduce((m, e) => e > m ? e : m);
export const min = <T>(...args: T[]) => args.reduce((m, e) => e < m ? e : m);

export const toBase64 = (arr: Uint8Array) => isNode ? Buffer.from(arr).toString('base64') : window.btoa(String.fromCharCode(...arr));
export const fromBase64 = (base64: string) => isNode ? new Uint8Array(Buffer.from(base64, 'base64')) : Uint8Array.from(atob(base64), c => c.charCodeAt(0))
