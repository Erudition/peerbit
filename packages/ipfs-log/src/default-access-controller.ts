import { Identities, IdentitySerializable } from "@dao-xyz/orbit-db-identity-provider";
import { Payload } from '@dao-xyz/ipfs-log-entry';

export interface AccessController<T> {
  canAppend(payload: Payload<T>, identityResolver: () => Promise<IdentitySerializable>, identityProvider: Identities): Promise<boolean> | boolean;
}

export class DefaultAccessController<T> implements AccessController<T> {
  async canAppend(_payload: Payload<T>, _identityResolver: () => Promise<IdentitySerializable>, identityProvider: Identities): Promise<boolean> {
    return true
  }
}
