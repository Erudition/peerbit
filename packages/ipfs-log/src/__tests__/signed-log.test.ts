const assert = require('assert')
const rmrf = require('rimraf')
const fs = require('fs-extra')
import { AccessController } from '../default-access-controller'
import { Log } from '../log'
import { Identities, Identity, IdentitySerializable } from '@dao-xyz/orbit-db-identity-provider'
import { assertPayload } from './utils/assert'
import { Keystore } from '@dao-xyz/orbit-db-keystore'
import { Entry } from '@dao-xyz/ipfs-log-entry'

// Test utils
const {
  config,
  testAPIs,
  startIpfs,
  stopIpfs
} = require('orbit-db-test-utils')

let ipfsd, ipfs, testIdentity: Identity, testIdentity2: Identity

Object.keys(testAPIs).forEach((IPFS) => {
  describe('Signed Log', function () {
    jest.setTimeout(config.timeout)

    const { identityKeyFixtures, signingKeyFixtures, identityKeysPath, signingKeysPath } = config

    let keystore, signingKeystore

    beforeAll(async () => {
      rmrf.sync(identityKeysPath)
      rmrf.sync(signingKeysPath)
      await fs.copy(identityKeyFixtures(__dirname), identityKeysPath)
      await fs.copy(signingKeyFixtures(__dirname), signingKeysPath)

      keystore = new Keystore(identityKeysPath)
      signingKeystore = new Keystore(signingKeysPath)


      testIdentity = await Identities.createIdentity({ id: new Uint8Array([0]), keystore, signingKeystore })
      testIdentity2 = await Identities.createIdentity({ id: new Uint8Array([1]), keystore, signingKeystore })
      ipfsd = await startIpfs(IPFS, config.defaultIpfsConfig)
      ipfs = ipfsd.api
    })

    afterAll(async () => {
      await stopIpfs(ipfsd)
      rmrf.sync(identityKeysPath)
      rmrf.sync(signingKeysPath)
      await keystore?.close()
      await signingKeystore?.close()
    })

    it('creates a signed log', () => {
      const logId = 'A'
      const log = new Log(ipfs, testIdentity, { logId })
      assert.notStrictEqual(log.id, null)
      assert.strictEqual(log.id, logId)
    })

    it('has the correct identity', () => {
      const log = new Log(ipfs, testIdentity, { logId: 'A' })
      assert.notStrictEqual(log.id, null)
      expect(log._identity.toSerializable()).toMatchSnapshot('identity');
    })

    it('has the correct public key', () => {
      const log = new Log(ipfs, testIdentity, { logId: 'A' })
      assert.strictEqual(log._identity.publicKey, testIdentity.publicKey)
    })

    it('has the correct pkSignature', () => {
      const log = new Log(ipfs, testIdentity, { logId: 'A' })
      assert.strictEqual(log._identity.signatures.id, testIdentity.signatures.id)
    })

    it('has the correct signature', () => {
      const log = new Log(ipfs, testIdentity, { logId: 'A' })
      assert.strictEqual(log._identity.signatures.publicKey, testIdentity.signatures.publicKey)
    })

    it('entries contain an identity', async () => {
      const log = new Log(ipfs, testIdentity, { logId: 'A' })
      await log.append('one')
      assert.notStrictEqual(await log.values[0].signature, null)
      assert.deepStrictEqual(await log.values[0].identity, testIdentity.toSerializable())
    })

    it('doesn\'t sign entries when identity is not defined', async () => {
      let err
      try {
        const log = new Log(ipfs, undefined, undefined) // eslint-disable-line no-unused-vars
      } catch (e) {
        err = e
      }
      assert.strictEqual(err.message, 'Identity is required')
    })

    it('doesn\'t join logs with different IDs ', async () => {
      const log1 = new Log<string>(ipfs, testIdentity, { logId: 'A' })
      const log2 = new Log<string>(ipfs, testIdentity2, { logId: 'B' })

      let err
      try {
        await log1.append('one')
        await log2.append('two')
        await log2.append('three')
        await log1.join(log2)
      } catch (e) {
        err = e.toString()
        throw e
      }

      assert.strictEqual(err, undefined)
      assert.strictEqual(log1.id, 'A')
      assert.strictEqual(log1.values.length, 1)
      assertPayload(log1.values[0].payload.value, 'one')
    })



    it('throws an error if log is signed but trying to merge an entry that doesn\'t have a signature', async () => {
      const log1 = new Log<string>(ipfs, testIdentity, { logId: 'A' })
      const log2 = new Log<string>(ipfs, testIdentity2, { logId: 'A' })

      let err
      try {
        await log1.append('one')
        await log2.append('two')
        delete (log2.values[0]._signature)
        await log1.join(log2)
      } catch (e) {
        err = e.toString()
      }
      assert.strictEqual(err, 'Error: Unsupported')
    })

    it('throws an error if log is signed but the signature doesn\'t verify', async () => {
      const log1 = new Log<string>(ipfs, testIdentity, { logId: 'A' })
      const log2 = new Log<string>(ipfs, testIdentity2, { logId: 'A' })
      let err

      try {
        await log1.append('one');
        await log2.append('two');
        let entry: Entry<string> = log2.values[0]
        entry._signature = await log1.values[0]._signature;
        await log1.join(log2)
      } catch (e) {
        err = e.toString()
      }

      const entry = log2.values[0]
      assert.strictEqual(err, `Error: Could not validate signature "${await entry.signature}" for entry "${entry.hash}" and key "${(await entry.identity).publicKey}"`)
      assert.strictEqual(log1.values.length, 1)
      assertPayload(log1.values[0].payload.value, 'one')
    })

    it('throws an error if entry doesn\'t have append access', async () => {
      const denyAccess = { canAppend: (_, __) => false } as AccessController<string>
      const log1 = new Log(ipfs, testIdentity, { logId: 'A' })
      const log2 = new Log(ipfs, testIdentity2, { logId: 'A', access: denyAccess })

      let err
      try {
        await log1.append('one')
        await log2.append('two')
        await log1.join(log2)
      } catch (e) {
        err = e.toString()
      }

      assert.strictEqual(err, `Error: Could not append Entry<T>, key "${testIdentity2.id}" is not allowed to write to the log`)
    })

    it('throws an error upon join if entry doesn\'t have append access', async () => {
      const testACL = {
        canAppend: async (_entry, identity, _) => Buffer.compare(Buffer.from(identity.decrypted.getValue(IdentitySerializable).id), Buffer.from(testIdentity.id)) === 0
      } as AccessController<string>;
      const log1 = new Log<string>(ipfs, testIdentity, { logId: 'A', access: testACL })
      const log2 = new Log<string>(ipfs, testIdentity2, { logId: 'A' })

      let err
      try {
        await log1.append('one')
        await log2.append('two')
        await log1.join(log2)
      } catch (e) {
        err = e.toString()
      }

      assert.strictEqual(err, `Error: Could not append Entry<T>, key "${testIdentity2.id}" is not allowed to write to the log`)
    })
  })
})
