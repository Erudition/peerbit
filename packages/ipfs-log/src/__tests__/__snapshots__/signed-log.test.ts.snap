// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signed Log has the correct identity: identity 1`] = `
IdentitySerializable {
  "id": Uint8Array [
    29,
    129,
    142,
    75,
    119,
    63,
    241,
    128,
    151,
    231,
    209,
    35,
    176,
    226,
    2,
    212,
    211,
    244,
    170,
    176,
    174,
    92,
    223,
    241,
    48,
    13,
    88,
    45,
    195,
    10,
    123,
    126,
  ],
  "publicKey": Ed25519PublicKey {
    "keyType": "ed25519",
    "publicKey": true,
  },
  "signatures": Signatures {
    "id": Uint8Array [
      119,
      49,
      16,
      128,
      251,
      46,
      0,
      254,
      21,
      119,
      63,
      26,
      4,
      3,
      163,
      105,
      94,
      28,
      252,
      106,
      154,
      251,
      212,
      134,
      31,
      23,
      114,
      87,
      39,
      96,
      130,
      87,
      2,
      164,
      100,
      80,
      191,
      107,
      163,
      73,
      61,
      146,
      238,
      227,
      227,
      211,
      218,
      108,
      47,
      56,
      241,
      86,
      58,
      78,
      173,
      174,
      206,
      227,
      47,
      136,
      210,
      243,
      30,
      7,
      29,
      129,
      142,
      75,
      119,
      63,
      241,
      128,
      151,
      231,
      209,
      35,
      176,
      226,
      2,
      212,
      211,
      244,
      170,
      176,
      174,
      92,
      223,
      241,
      48,
      13,
      88,
      45,
      195,
      10,
      123,
      126,
    ],
    "publicKey": Uint8Array [
      54,
      81,
      244,
      208,
      78,
      36,
      34,
      60,
      39,
      87,
      111,
      237,
      24,
      230,
      197,
      177,
      197,
      114,
      59,
      23,
      229,
      61,
      107,
      127,
      112,
      37,
      53,
      58,
      86,
      98,
      119,
      118,
      94,
      161,
      131,
      99,
      16,
      50,
      115,
      246,
      74,
      185,
      242,
      186,
      198,
      92,
      89,
      10,
      227,
      240,
      107,
      215,
      150,
      31,
      141,
      139,
      212,
      120,
      31,
      107,
      53,
      24,
      173,
      2,
      223,
      92,
      207,
      196,
      251,
      189,
      159,
      245,
      23,
      2,
      12,
      208,
      80,
      121,
      235,
      200,
      242,
      60,
      215,
      57,
      91,
      169,
      234,
      200,
      168,
      137,
      238,
      168,
      123,
      8,
      47,
      113,
      119,
      49,
      16,
      128,
      251,
      46,
      0,
      254,
      21,
      119,
      63,
      26,
      4,
      3,
      163,
      105,
      94,
      28,
      252,
      106,
      154,
      251,
      212,
      134,
      31,
      23,
      114,
      87,
      39,
      96,
      130,
      87,
      2,
      164,
      100,
      80,
      191,
      107,
      163,
      73,
      61,
      146,
      238,
      227,
      227,
      211,
      218,
      108,
      47,
      56,
      241,
      86,
      58,
      78,
      173,
      174,
      206,
      227,
      47,
      136,
      210,
      243,
      30,
      7,
      29,
      129,
      142,
      75,
      119,
      63,
      241,
      128,
      151,
      231,
      209,
      35,
      176,
      226,
      2,
      212,
      211,
      244,
      170,
      176,
      174,
      92,
      223,
      241,
      48,
      13,
      88,
      45,
      195,
      10,
      123,
      126,
    ],
  },
  "type": "orbitdb",
}
`;
